OpenRA.Mods.Common/Scripting/Properties/DockClientManagerProperties.cs

```diff
@@ -0,0 +1,88 @@
+#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using Eluant;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Primitives;
+using OpenRA.Scripting;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Scripting
+{
+	[ScriptPropertyGroup("Ability")]
+	public class DockClientManagerProperties : ScriptActorProperties, Requires<DockClientBaseInfo>
+	{
+		readonly DockClientManager manager;
+
+		public DockClientManagerProperties(ScriptContext context, Actor self)
+			: base(context, self)
+		{
+			manager = self.Trait<DockClientManager>();
+		}
+
+		[ScriptActorPropertyActivity]
+		[Desc("Can we dock to this actor?")]
+		public bool CanDockAt(Actor target, bool forceEnter = false, bool ignoreOccupancy = false)
+		{
+			if (target == null)
+				throw new LuaException("Target actor is null.");
+
+			return manager.CanDockAt(target, forceEnter, ignoreOccupancy);
+		}
+
+		[ScriptActorPropertyActivity]
+		[Desc("Dock to the actor. Returns true if docking activity was successfully queued.")]
+		public bool Dock(Actor target, bool forceEnter = false, bool ignoreOccupancy = true)
+		{
+			if (target == null)
+				throw new LuaException("Target actor is null.");
+
+			if (manager.IsTraitDisabled)
+				return false;
+
+			var docks = manager.AvailableDockHosts(target, forceEnter, ignoreOccupancy).ToList();
+			if (docks.Count == 0)
+				return false;
+
+			var dock = docks.ClosestDock(Self, manager);
+			if (!dock.HasValue)
+				return false;
+
+			Self.QueueActivity(new MoveToDock(Self, dock.Value.Actor, dock.Value.Trait));
```

Comment:
Why isn't there an activity that takes a `targetActor` and does the above calculations as soon as the actor has moved close enough to the target? The target actor/dock might have changed while moving, and we don't explicitly request a dock type here so don't care about any changes while moving to the target.

Reply:
What is the limiting property?

Reply:
in C# you are supposed to just pass in the relevant IDockHost class / itnerface

Reply:
That sounds weird, aren't there dock types?

Reply:
you are supposed to get valid docks by dock type (and other factors), and then dock to the one of your choosing

Reply:
The Lua part here gets the target actor passed, so I'm still not sure why we can't do the necessary calculations once at the target. 🤔

Reply:
This method might be too generic. I'll have to think about the use-cases.

Reply:
The player cursor order also passes in Actor as target, just as this lua function. And as for choosing valid docks at the destination, remember that they can have offsets. Some may be unreachable or require a completely different path